src/QsCompiler/Compiler/CompilationLoader.cs

```diff
@@ -568,14 +573,15 @@ public CompilationLoader(SourceLoader loadSources, ReferenceLoader loadReference
             // executing the specified rewrite steps
             PerformanceTracking.TaskStart(PerformanceTracking.Task.RewriteSteps);
             var steps = new List<(int, string, Func<QsCompilation?>)>();
-            var qirEmissionEnabled = this.externalRewriteSteps.Any(step => step.Name == "QIR Generation");
-            if (this.config.IsExecutable && !this.config.SkipSyntaxTreeTrimming)
+            this.config.IsQirGenerationEnabled = this.externalRewriteSteps.Any(step => step.Name == "QIR Generation");
+
+            if ((this.config.IsExecutable || this.config.IsQirGenerationEnabled) && !this.config.SkipSyntaxTreeTrimming)
             {
                 // TODO: It would be nicer to trim unused intrinsics. Currently, this is not possible due to how the old setup of the C# runtime works.
                 // With the new setup (interface-based approach for target packages), it is possible to trim ununsed intrinsics.
 #pragma warning disable CS0618 // Type or member is obsolete
                 // TODO: The dependencies for rewrite steps should be declared as part of IRewriteStep interface, and we should query those here.
-                var rewriteStep = new LoadedStep(new SyntaxTreeTrimming(keepAllIntrinsics: true, BuiltIn.AllBuiltIns.Select(e => e.FullName)), typeof(IRewriteStep), thisDllUri);
+                var rewriteStep = new LoadedStep(new SyntaxTreeTrimming(keepAllIntrinsics: !this.config.IsQirGenerationEnabled, BuiltIn.AllBuiltIns.Select(e => e.FullName), !this.config.IsExecutable), typeof(IRewriteStep), thisDllUri);

```

Comment:
@cesarzc @ScottCarda-MS I noticed this during merge conflict resolution and was wondering if you could explain the reason for hard-coding true here. In particular, I thought preserving intrinsics should not be necessary when generating QIR because it doesn't rely on any inheritance. Is there a new need to preserve all intrinsics even for QIR because of how some of them get used?

Reply:
@bettinaheim first introduced the keepAllIntrinsics. She may have a better memory on this than me.

Reply:
What callables (including intrinsics) should be kept and which ones should be trimmed should be determined based on usage (see the above TODOs), so setting this to true may have just been a temporary approach until we have a more robust dependency-based logic here.